src/merkle/smt/full/concurrent/mod.rs

```diff
@@ -255,28 +257,30 @@ impl Smt {
     /// # Panics
     /// With debug assertions on, this function panics if it detects that `pairs` is not correctly
     /// sorted. Without debug assertions, the returned computations will be incorrect.
-    fn sorted_pairs_to_leaves(pairs: Vec<(RpoDigest, Word)>) -> PairComputations<u64, SmtLeaf> {
-        Self::process_sorted_pairs_to_leaves(pairs, |leaf_pairs| {
-            Some(Self::pairs_to_leaf(leaf_pairs))
-        })
+    fn sorted_pairs_to_leaves(
+        pairs: Vec<(RpoDigest, Word)>,
+    ) -> Result<PairComputations<u64, SmtLeaf>, MerkleError> {
+        Self::process_sorted_pairs_to_leaves(pairs, Self::pairs_to_leaf)
     }
 
     /// Constructs a single leaf from an arbitrary amount of key-value pairs.
     /// Those pairs must all have the same leaf index.
-    fn pairs_to_leaf(mut pairs: Vec<(RpoDigest, Word)>) -> SmtLeaf {
+    fn pairs_to_leaf(mut pairs: Vec<(RpoDigest, Word)>) -> Result<Option<SmtLeaf>, MerkleError> {
         assert!(!pairs.is_empty());
 
         if pairs.len() > 1 {
             pairs.sort_by(|(key_1, _), (key_2, _)| leaf::cmp_keys(*key_1, *key_2));
-            SmtLeaf::new_multiple(pairs).unwrap()
+            if let Some(window) = pairs.windows(2).find(|w| w[0].0 == w[1].0) {
```

Comment:
nit: I'd add a comment that you are searching for duplicates in a sorted array. I know it's already in the context (`sort_by`, `DuplicateValues`) but it's a bit tough to read with everything being named `w`, `[0]`, and `.0`.